File: hmis-engine-api/src/main/java/com/otapp/hmis/engine/masterdata/dosage/application/DosageService.java
package com.otapp.hmis.engine.masterdata.dosage.application;

import com.otapp.hmis.engine.common.api.PageResponse;
import com.otapp.hmis.engine.common.error.ConflictException;
import com.otapp.hmis.engine.common.error.NotFoundException;
import com.otapp.hmis.engine.masterdata.dosage.application.DosageDtos.CreateDosageRequest;
import com.otapp.hmis.engine.masterdata.dosage.application.DosageDtos.DosageDto;
import com.otapp.hmis.engine.masterdata.dosage.application.DosageDtos.UpdateDosageRequest;
import com.otapp.hmis.engine.masterdata.dosage.domain.Dosage;
import com.otapp.hmis.engine.masterdata.dosage.domain.DosageRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class DosageService {

    private final DosageRepository repo;

    @Transactional
    public DosageDto create(CreateDosageRequest request) {
        String code = request.code().trim().toUpperCase();
        if (repo.existsByCode(code)) {
            throw new ConflictException("Dosage code already exists: " + code);
        }
        Dosage d = new Dosage(code, request.name().trim(), request.description());
        repo.save(d);
        return toDto(d);
    }

    @Transactional
    public DosageDto update(String uid, UpdateDosageRequest request) {
        Dosage d = loadOrThrow(uid);
        d.setName(request.name().trim());
        d.setDescription(request.description());
        return toDto(d);
    }

    @Transactional
    public DosageDto setActive(String uid, boolean active) {
        Dosage d = loadOrThrow(uid);
        if (active) d.activate(); else d.deactivate();
        return toDto(d);
    }

    @Transactional
    public void delete(String uid) { repo.delete(loadOrThrow(uid)); }

    @Transactional(readOnly = true)
    public DosageDto findByUid(String uid) { return toDto(loadOrThrow(uid)); }

    @Transactional(readOnly = true)
    public PageResponse<DosageDto> search(String query, Boolean active, Pageable pageable) {
        return PageResponse.from(
                repo.search(query == null ? null : query.trim(), active, pageable)
                        .map(DosageService::toDto));
    }

    private Dosage loadOrThrow(String uid) {
        return repo.findByUid(uid).orElseThrow(() -> new NotFoundException("Dosage not found: " + uid));
    }

    private static DosageDto toDto(Dosage d) {
        return new DosageDto(d.getUid(), d.getCode(), d.getName(), d.getDescription(),
                d.isActive(), d.getCreatedAt(), d.getUpdatedAt());
    }
}

File: hmis-engine-web/src/app/features/masterdata/dosages/dosage.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../../environments/environment';
import { PageResponse } from '../../../core/http/page.types';
import {
  CreateDosageRequest, Dosage, DosageSearchParams, UpdateDosageRequest
} from './dosage.types';

@Injectable({ providedIn: 'root' })
export class DosageService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/masterdata/dosages`;

  search(params: DosageSearchParams = {}): Observable<PageResponse<Dosage>> {
    let p = new HttpParams();
    if (params.query) p = p.set('query', params.query);
    if (params.active !== undefined) p = p.set('active', String(params.active));
    if (params.page !== undefined) p = p.set('page', String(params.page));
    if (params.size !== undefined) p = p.set('size', String(params.size));
    if (params.sort) p = p.set('sort', params.sort);
    return this.http.get<PageResponse<Dosage>>(this.base, { params: p });
  }

  findByUid(uid: string): Observable<Dosage> { return this.http.get<Dosage>(`${this.base}/uid/${uid}`); }
  create(req: CreateDosageRequest): Observable<Dosage> { return this.http.post<Dosage>(this.base, req); }
  update(uid: string, req: UpdateDosageRequest): Observable<Dosage> { return this.http.put<Dosage>(`${this.base}/uid/${uid}`, req); }
  setActive(uid: string, active: boolean): Observable<Dosage> { return this.http.put<Dosage>(`${this.base}/uid/${uid}/active`, { active }); }
  delete(uid: string): Observable<void> { return this.http.delete<void>(`${this.base}/uid/${uid}`); }
}
